### PR TITLE
expose dpiConn_newVar

### LINKED ISSUE #134: Need help for calling PL/SQL Stored Proc with custom object as input
Not able to call an Oracle stored procedure which takes a custom object as input. Sample here:
PROCEDURE set_test_comments(
      p_test_comments_i    IN TEST_COMMENTS_OBJECT,
      p_state_o                  OUT varchar2)

CREATE OR REPLACE TYPE  TEST_O.TEST_COMMENTS_OBJECT FORCE AS OBJECT
(user_id  NUMBER,
user_name   VARCHAR2(20),
comment_date   DATE,
comments   VARCHAR2(4000)
)

Tried some options but none of them helped, so would really appreciate any suggestions on how to move forward! Don't have a running code yet.
1. Get the object type, and then create a new object. But unable to set the input data - 
objType, err := goracle.GetObjectType(tx, "TEST_O.TEST_COMMENTS_OBJECT")
inputObj, err := objType.NewObject()
2. Set each attribute in the above object by creating a Data variable and set bytes/bool (depending on input data type) but end up getting stack overflow error:
var tmp goracle.Data
tmp.SetBool(true)
inputObj.SetAttribute("COMMENTS", &tmp)

### PR DESCRIPTION
**Pull request by wwanderley**
Create new goracle.Data struct to use as input parameter on ObjectCollection
Close #134

### CODE REVIEW

File: data.go [hunk 1]
```diff
@@ -283,6 +287,109 @@ func (d *Data) IsObject() bool {
...
+	}
+
+	var natType C.dpiNativeTypeNum
+	var oraType C.dpiOracleTypeNum
+	var objectType *C.dpiObjectType
+	var obj ObjectType
+
+	switch v := baseType.(type) {
+	case Lob, []Lob:
+		natType = C.DPI_NATIVE_TYPE_LOB
```
Comment (tgulacsi) [inline data.go]: case []Lob:	
                isSlice = True
                fallthrough
case Lob: 
		natType = C.DPI_NATIVE_TYPE_LOB
		var isClob bool
		switch v := v.(type) {
		case Lob:
			isClob = v.IsClob
		case []Lob:
			isClob = len(v) > 0 && v[0].IsClob
		}		
		if isClob {
			oraType = C.DPI_ORACLE_TYPE_CLOB
		} else {
			oraType = C.DPI_ORACLE_TYPE_BLOB
		}

File: data.go [hunk 2]
```diff
@@ -283,6 +287,109 @@ func (d *Data) IsObject() bool {
...
+			if len(v) > 0 {
+				objectType = v[0].ObjectRef().ObjectType.dpiObjectType
+			}
+		}		
+	default:		
+		return nil, errors.Errorf("unknown type %T", v)
+	}
+
+	isArray := C.int(0)
+	if reflect.TypeOf(baseType).Kind() == reflect.Slice {
```
Comment (tgulacsi) [inline data.go]: No need for reflect if we "fallthrough" the slice cases in the type switch and set isSlice there.

File: data.go [hunk 3]
```diff
@@ -283,6 +287,109 @@ func (d *Data) IsObject() bool {
...
+		isArray = 1
+	}
+	if sliceLen < 1 {
+		sliceLen = 1
+	}
+	var dataArr *C.dpiData
+	var v *C.dpiVar
```
Comment (tgulacsi) [inline data.go]: Maybe use conn.newVar here? Seems to do the same.

Maybe split out the upper type switch into a NewVarInfo which creates a varInfo (maybe export it),
and make NewDataAny just a comfort wrapper of NewVarInfo + NewVar + NewData,
and NewData gets a VarInfo , and so on.

The goal is to have usable building blocks, not smart and complex functions.

File: data.go [hunk 4]
```diff
@@ -283,6 +287,109 @@ func (d *Data) IsObject() bool {
...
+	}
+	if C.dpiConn_newVar(
+		c.dpiConn, oraType, natType, C.uint32_t(sliceLen),
+		C.uint32_t(bufSize), 1,
+		isArray, objectType,
+		&v, &dataArr,
+	) == C.DPI_FAILURE {
+		return nil, errors.Wrapf(c.getError(), "newVar(typ=%d, natTyp=%d, sliceLen=%d, bufSize=%d)", oraType, natType, sliceLen, bufSize)
+	}
+	C.dpiVar_release(v)
```
Comment (tgulacsi) [inline data.go]: What is the dpiConn_newVar does if we promptly release it ?
As I see in odpi/src/dpiVar.c, the returned dpiData is a member of the dpiVar struct, so I'm afraid if we release the dpiVar, we lose the dpiData, too.

File: data.go (hunk 1 above)
Comment (wwanderley) [reply data.go]: Is it possible to "fallthrough"  in a type switch?

File: data.go (hunk 2 above)
Comment (wwanderley) [reply data.go]: "cannot fallthrough in type switch"

File: data.go (hunk 3 above)
Comment (wwanderley) [reply data.go]: Nice review. Thank you!

- I use conn.newVar now. 
- Keep varInfo unexported (it has ODPI-C types that must be protected) 
- Create func newVarInfo as suggested

File: data.go (hunk 4 above)
Comment (wwanderley) [reply data.go]: Ok, i will remove this.

File: data.go (hunk 1 above)
Comment (tgulacsi) [reply data.go]: You're right, type switch is not a regular switch, forgot it.

**Pull request merged**